pkg/controllers/dnspolicy/health.go

```diff
@@ -0,0 +1,217 @@
+			return err
+		}
+	}
+
+	if err = r.updateDNSRecord(ctx, dnsRecordOriginal, dnsRecord); err != nil {
+		return err
+	}
+
+	return r.reconcileHealthCheckStatus(results, policy)
+}
+
+func (r *DNSPolicyReconciler) reconcileHealthChecks(ctx context.Context, log logr.Logger, dnsRecord *v1alpha1.DNSRecord, policy *v1alpha1.DNSPolicy, config *healthChecksConfig) ([]dns.HealthCheckResult, error) {
+	healthCheckReconciler := r.DNSProvider.HealthCheckReconciler()
+
+	results := []dns.HealthCheckResult{}
+
+	for _, dnsEndpoint := range dnsRecord.Spec.Endpoints {
+		ok := false
+		if _, ok = dnsEndpoint.GetAddress(); !ok {
+			// TODO: Log
+			continue
+		}
+
+		endpointId, err := idForEndpoint(dnsRecord, dnsEndpoint)
+		if err != nil {
+			return nil, err
+		}
+
+		spec := dns.HealthCheckSpec{
+			Id:               endpointId,
+			Name:             fmt.Sprintf("%s-%s", dnsEndpoint.DNSName, dnsEndpoint.SetIdentifier),
+			Path:             config.Endpoint,
+			Port:             config.Port,
+			Protocol:         config.Protocol,
+			FailureThreshold: config.FailureThreshold,
+		}
+
+		result, err := healthCheckReconciler.Reconcile(ctx, spec, dnsEndpoint)
+		if err != nil {
+			return nil, err
+		}
+
+		results = append(results, result)
+	}
+
+	return results, nil
+}
+
+func (r *DNSPolicyReconciler) reconcileHealthCheckStatus(results []dns.HealthCheckResult, policy *v1alpha1.DNSPolicy) error {
+	for _, result := range results {
+		if result.Result == dns.HealthCheckNoop {
+			continue
+		}
+
+		if policy.Status.HealthCheck == nil {
+			policy.Status.HealthCheck = &v1alpha1.HealthCheckStatus{}
+		}
+
+		status := metav1.ConditionTrue
+		if result.Result == dns.HealthCheckFailed {
+			status = metav1.ConditionFalse
+		}
+
+		policy.Status.HealthCheck.Conditions = append(policy.Status.HealthCheck.Conditions, metav1.Condition{
+			ObservedGeneration: policy.Generation,
+			Status:             status,
+			Reason:             string(result.Result),
+			LastTransitionTime: metav1.Now(),
+			Message:            result.Message,
+			Type:               string(result.Result),
+		})
+	}
+
+	return nil
+}
+
+func (r *DNSPolicyReconciler) reconcileHealthCheckDeletion(ctx context.Context, dnsRecord *v1alpha1.DNSRecord, policy *v1alpha1.DNSPolicy) ([]dns.HealthCheckResult, error) {
+	reconciler := r.DNSProvider.HealthCheckReconciler()
+	results := []dns.HealthCheckResult{}
+
+	for _, endpoint := range dnsRecord.Status.Endpoints {
+		result, err := reconciler.Delete(ctx, endpoint)
+		if err != nil {
+			return nil, err
+		}
+
+		results = append(results, result)
+	}
+
+	return results, nil
+}
+
+func (r *DNSPolicyReconciler) getDNSRecord(ctx context.Context, policy *v1alpha1.DNSPolicy) (*v1alpha1.DNSRecord, bool, error) {
+	gateway, err := r.getGateway(ctx, policy)
+	if err != nil {
+		return nil, false, err
+	}
+
+	dnsRecordList := &v1alpha1.DNSRecordList{}
+	if err := r.Client.List(ctx, dnsRecordList); err != nil {
```

Comment:
does this list all records? I think we could def limit it to the same namespace as the gateway?

Reply:
actually not seeing where dnsRecordList is used ?

Reply:
Oops, that's actually a bit that I forgot to remove after I changed the implementation to get the DNSRecord